apollo-router/src/services/execution.rs

```diff
@@ -66,3 +103,35 @@ impl Request {
 
 /// The response type for execution services is the same as for supergraph services.
 pub type Response = super::supergraph::Response;
+
+// Even though execution response is now almost exactly the same as supergraph Response, we need a
+// mechanism to propagate headers from the subgraph, so we provide this extra function here.
+impl Response {
+    /// This is the constructor to use when constructing a real ExecutionResponse.
+    ///
+    /// In this case, you already have a valid request and just wish to associate it with a context
+    /// and create a ExecutionResponse.
+    pub(crate) fn new_from_response(
+        mut response: http::Response<BoxStream<'static, graphql::Response>>,
+        headers_opt: Option<HeaderMap<HeaderValue>>,
+        context: Context,
+    ) -> Self {
+        if let Some(headers) = headers_opt {
+            headers
+                .into_iter()
+                .filter(|(name_opt, _)| {
+                    let name = name_opt.as_ref().expect("name must be valid");
+                    !RESERVED_HEADERS.contains(name)
+                })
+                .for_each(|(name_opt, value)| {
+                    let name = name_opt.expect("name must be valid");
+                    tracing::info!("inserting header: {}", name);
+                    response.headers_mut().insert(name, value);
```

Comment:
the `into_iter()` here looks like it will do the expected behavior of iterating over _all_ headers in the response, including repeated ones.

But... `insert` here will [overwrite](https://docs.rs/http/0.1.3/http/struct.HeaderMap.html#method.insert) repeated headers with whatever the last instance of that header contained. I would suggest using `append` instead, as [documented here](https://docs.rs/http/0.1.3/http/struct.HeaderMap.html#method.append).

If you consider the case of a subgraph that needs to set multiple cookies on the client browser, it will respond with multiple `Set-Cookie` headers containing different values. The code as written looks like it will randomly drop all cookies but the final one, which seems like incorrect behavior.

Reply:
Thanks for the review!
Yes. It's a PofC which needs some (a lot of) polishing yet. The purpose of this draft was to demonstrate that we could get headers back from subgraphs to clients.
We now need to position this basic mechanism (and fix bugs such as you describe) within a framework that we can expose via configuration (as described in #1284). The tricky bit will be accessing the configuration within the header propagation logic to make decisions such as:
 - which subgraphs should we propagate which headers from?
 - should we insert headers for certain subgraphs?
 - etc...

Our intent is to try and apply something like the basic framework for header manipulation which we already have for requests to responses.